src/cloudflare/internal/pipeline-transform.ts

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2024 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+import entrypoints from 'cloudflare-internal:workers'
+
+async function* readLines(stream: ReadableStream<string>): AsyncGenerator<string> {
+  let partial = ''
+
+  // @ts-ignore
+  for await (const chunk of stream) {
+    const full = partial + chunk
+    const lines = full.split('\n')
+    const popped = lines.pop()
+    if (popped) {
+      partial = popped
+    }
+
+    for (const line of lines) {
+      yield line;
+    }
+  }
+
+  if (partial.length > 0) {
+    yield partial;
+  }
+}
+
+type Batch = {
+  id: string 	// unique identifier for the batch
+  shard?: string 	// assigned shard
+  ts: number 	// timestamp of the event
+
+  format: Format
+  data: unknown
+}
+
+type JsonStream = Batch & {
+  format: Format.JSON_STREAM,
+  data: ReadableStream<Uint8Array>
+}
+
+enum Format {
+  JSON_STREAM = 'json_stream', //jsonl
+}
+
+export class PipelineTransformImpl extends entrypoints.WorkerEntrypoint {
+  #batch?: Batch
+  #initalized: boolean = false
+
+  constructor(ctx: unknown, env: unknown) {
+    super(ctx, env)
+  }
+
+  // stub overriden on the sub class
+  // @ts-ignore
+  public async transformJson(data: object[]): Promise<object[]> {
+    throw new Error('should be implemented by parent')
+  }
+
+  // called by the dispatcher which then calls the subclass methods
+  // @ts-ignore
+  private async _transform(batch: Batch, ping?: boolean): Promise<JsonStream> {
+    if (this.#initalized) {
+      throw new Error('pipeline entrypoint has already been initialized')
```

Comment:
Would we ever need to reset/restart a stream for reliability purposes?  Seems like this should handle either case of multiple streams or single infinite stream.